flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/SocketTextStreamFunction.java

```diff
@@ -70,14 +74,15 @@ public void run(SourceContext<String> ctx) throws Exception {
 
 	private void streamFromSocket(SourceContext<String> ctx, Socket socket) throws Exception {
 		try {
-			StringBuilder buffer = new StringBuilder();
+			StringBuffer buffer = new StringBuffer();
+			char[] charBuffer = new char[Math.max(8192, 2 * delimiter.length())];
 			BufferedReader reader = new BufferedReader(new InputStreamReader(
 					socket.getInputStream()));
 
 			while (isRunning) {
-				int data;
+				int readCount;
 				try {
-					data = reader.read();
+					readCount = reader.read(charBuffer);
```

Comment:
Good change. It's much better to read a buffer instead of individual characters.